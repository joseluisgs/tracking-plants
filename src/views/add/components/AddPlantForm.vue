<template>
  <ValidationObserver v-slot="{ invalid }">
    <form class="mt-8" @submit.prevent="onSubmit">
      <!-- Recibe el evento llamado input de autocomplete
      Validacion con Validate
      Componente personalizado -->
      <ValidationProvider
        name="autocomplete"
        rules="required"
        v-slot="{ errors }"
      >
        <label class="font-bold" for="autocomplete">Busca una planta</label>
        <Autocomplete
          v-model="autocompleteValue"
          :items="plants"
          placeholder="Nombre de la Planta"
          @is-typing="onChangeAutocomplete"
          @select-item="onSelectItem"
          @clear-items="clearItems"
        />
        <span>{{ errors[0] }}</span>
      </ValidationProvider>
      <div class="flex justify-center mt-8">
        <!-- Cambio a componentes personalizado -->
        <TheButton size="large" :is-disabled="invalid"> Guardar </TheButton>
      </div>
    </form>
  </ValidationObserver>
</template>

<script>
import Autocomplete from '@/components/ui/Autocomplete.vue';
import TheButton from '@/components/ui/TheButton.vue';
import { ValidationProvider, ValidationObserver } from 'vee-validate';

export default {
  name: 'AddPlantForm',
  // Mis componentes
  components: {
    Autocomplete,
    TheButton,
    ValidationProvider,
    ValidationObserver,
  },
  // Mi modelo de datos
  data: () => ({
    currentPlant: null,
    autocompleteValue: null,
  }),
  // Mis propiedades, me llegan desde el esterior
  props: {
    plants: {
      type: Array,
      default: () => [],
    },
  },
  // Mis mÃ©todos
  methods: {
    async onChangeAutocomplete(value) {
      if (value.length >= 3) {
        this.$emit('find-plant', value);
        return;
      }
      this.clearItems();
    },
    onSelectItem(item) {
      this.autocompleteValue = item.name;
      this.currentPlant = item;
      this.clearItems();
      this.$emit('selected-plant', item);
    },
    clearItems() {
      this.$emit('clear-items');
    },
    onSubmit() {
      this.$emit('submit-form', this.currentPlant);
      this.clearAll();
    },
    clearAll() {
      this.clearItems();
      this.currentPlant = null;
      this.autocompleteValue = null;
    },
  },
};
</script>

<style>
</style>
