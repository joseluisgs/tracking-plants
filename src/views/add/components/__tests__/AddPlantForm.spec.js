import { shallowMount } from '@vue/test-utils';
import AddPlantForm from '../AddPlantForm.vue';

describe('Add View -> Component -> AddPlantForm', () => {
  // Nos devuelve un shallowmont del componente
  // es decir una copia de lo que se renderizarÃ­a en la web
  // Hacemos el build metiendole las propos, para ahorrarnos en cada uno hacer wrapper.setProps({ theme: 'primary' });
  const build = (propsData) => {
    const options = propsData ? { propsData } : null;
    const wrapper = shallowMount(AddPlantForm, options);

    return {
      wrapper,
    };
  };

  // Coicide su snapshot
  test('should match snapshot', () => {
    const { wrapper } = build();
    expect(wrapper.element).toMatchSnapshot();
  });

  // Probamos evento onChangeAutocomplete
  test('should emit onChangeAutocomplete', () => {
    const { wrapper } = build();
    const value = [1, 2, 3];
    wrapper.vm.onChangeAutocomplete(value);
    expect(wrapper.emitted()['find-plant']).toBeDefined();
    expect(wrapper.emitted()['find-plant']).toEqual([[value]]);
  });
});
