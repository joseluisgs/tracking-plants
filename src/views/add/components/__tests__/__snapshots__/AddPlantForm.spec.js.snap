// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Add View -> Component -> AddPlantForm should match snapshot 1`] = `
<validationobserver-stub
  tag="span"
  vid="obs_0"
>
  <form
    class="mt-8"
  >
    <validationprovider-stub
      bails="true"
      custommessages="[object Object]"
      debounce="0"
      detectinput="true"
      mode="aggressive"
      name="autocomplete"
      rules="required"
      skipifempty="true"
      tag="span"
      vid=""
    />
     
    <div
      class="flex justify-center mt-8"
    >
      <thebutton-stub
        size="large"
        theme="primary"
      >
         Guardar 
      </thebutton-stub>
    </div>
  </form>
</validationobserver-stub>
`;
